tests/components/screenlogic/test_config_flow.py

```diff
@@ -0,0 +1,249 @@
+        context={"source": "dhcp"},
+        data={
+            HOSTNAME: "Pentair: 01-01-01",
+            IP_ADDRESS: "1.1.1.1",
+        },
+    )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "gateway_entry"
+
+
+async def test_form_manual_entry(hass):
+    """Test we get the form."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    with patch(
+        "homeassistant.components.screenlogic.config_flow.discover",
+        return_value=[
+            {
+                SL_GATEWAY_IP: "1.1.1.1",
+                SL_GATEWAY_PORT: 80,
+                SL_GATEWAY_TYPE: 12,
+                SL_GATEWAY_SUBTYPE: 2,
+                SL_GATEWAY_NAME: "Pentair: 01-01-01",
+            },
+        ],
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+    assert result["type"] == "form"
+    assert result["errors"] == {}
+    assert result["step_id"] == "gateway_select"
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={GATEWAY_SELECT_KEY: GATEWAY_MANUAL_ENTRY}
+    )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {}
+    assert result2["step_id"] == "gateway_entry"
+
+    with patch(
+        "homeassistant.components.screenlogic.config_flow.login.create_socket",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.screenlogic.config_flow.login.gateway_connect",
+        return_value="00-C0-33-01-01-01",
+    ):
+        result3 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_IP_ADDRESS: "1.1.1.1",
+                CONF_PORT: 80,
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result3["type"] == "create_entry"
+    assert result3["title"] == "Pentair: 01-01-01"
+    assert result3["data"] == {
+        CONF_IP_ADDRESS: "1.1.1.1",
+        CONF_PORT: 80,
+    }
+
+
+async def test_form_cannot_connect(hass):
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.screenlogic.config_flow.login.create_socket",
+        return_value=None,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_IP_ADDRESS: "1.1.1.1",
+                CONF_PORT: 80,
+            },
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {CONF_IP_ADDRESS: "cannot_connect"}
+
+
+async def test_option_flow(hass):
+    """Test config flow options."""
+    entry = MockConfigEntry(domain=DOMAIN, data={}, options=None)
+    entry.add_to_hass(hass)
+
+    result = await hass.config_entries.options.async_init(entry.entry_id)
```

Comment:
The integration needs to be set up before we start the options flow to make sure that the config flow module is loaded. Use `hass.config_entries.async_setup`.

Reply:
I'm having trouble with this. I have 3 tests of the options flow and I have added
```
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()
```
to each, after adding the mock entry and before calling `async_init()`, as other integrations have done. The first one to run passes fine, but the remaining 2 fail in `async_setup()` with:
```
        if entry.state != ENTRY_STATE_NOT_LOADED:
>           raise OperationNotAllowed
E           homeassistant.config_entries.OperationNotAllowed
```
I'm not sure what I'm doing wrong.

Reply:
I think I resolved this. I was trying to test with a full `MockConfigEntry`, instead of one that only specifies the `domain`.